### DISCUSSION TITLE
PostGIS for PostgreSQL 15

### DISCUSSION THREAD

**Discussion by boris-glumpler** [Homebrew/discussions#3987] [Everyday usage]
It seems that the current version of PostGIS still requires v14 of Postgres. Is there a way to force it to install for v15? I tried to update the dependency with `brew edit postgis`. This seemed to work in that it did not install Postgres 14 and didn't throw any errors. Trying to then add the extension to a database doesn't work, tho. It says that PostGIS is not installed. Would greatly appreciate any pointers! Thanks!

**Comment by boris-glumpler** [comment]
I ended up compiling PostGIS from source against Postgres 15. Cheers!

**Reply by fbonzon** [reply]
@boris-glumpler

> I ended up compiling PostGIS from source against Postgres 15.

Can you share more details? Using a `brew` command?

I upgraded the PostgreSQL version too with `brew edit postgis` and executed `brew install --build-from-source postgis`, but got this error:

```
==> ./configure --with-projdir=/opt/homebrew/opt/proj --with-jsondir=/opt/homebrew/opt/json-c --with-pgconfig=/opt/homebrew/opt/postgresql@15/bin/pg_config -
==> make
==> make install DESTDIR=/private/tmp/postgis-20221129-26711-1pjztbx/postgis-3.3.2/stage
Error: An exception occurred within a child process:
  Errno::ENOENT: No such file or directory @ dir_initialize - /private/tmp/postgis-20221129-26711-1pjztbx/postgis-3.3.2/stage/opt/homebrew/lib
```

Did you do more changes in `postgis.rb` other than the 2 code occurrences of `postgresql@14` → `postgresql@15` ?

**Reply by boris-glumpler** [reply]
Nope, not using `brew`. Just followed the directions on https://postgis.net/docs/manual-3.3/postgis_installation.html. 

```
tar -xvfz postgis-3.3.2.tar.gz
cd postgis-3.3.2
./configure --with-projdir=/usr/local/opt/proj --with-jsondir=/usr/local/opt/json-c --with-pgconfig=/usr/local/opt/postgresql@15/bin/pg_config  --with-protobufdir=/usr/local/opt/protobuf-c  --disable-nls
make
make install
```

**Reply by svoop** [reply]
@boris-glumpler Have you figured out what caused the "No such file or directory" error?

**Reply by svoop** [reply]
This is where the error occurs, line 91 and 92 fail because no `lib` nor `share` are present:

https://github.com/Homebrew/homebrew-core/blob/09b5a198d63f2067d0e1b3f5cc5bedb53f37cefc/Formula/postgis.rb#L91

@SMillerDev Have you maybe seen someting like this before? (Sorry, I'm not much of a C-guy.)

---

```
  Errno::ENOENT: No such file or directory @ dir_initialize - /private/tmp/postgisA3.3.2-20221208-26426-18ns24s/postgis-3.3.2/stage/opt/homebrew/lib
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/pathname.rb:443:in `open'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/pathname.rb:443:in `foreach'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/pathname.rb:443:in `children'
/opt/homebrew/Library/Taps/sschwyn/homebrew-local-postgis/Formula/postgis@3.3.2.rb:81:in `install'
/opt/homebrew/Library/Homebrew/debrew.rb:17:in `block in install'
/opt/homebrew/Library/Homebrew/debrew.rb:90:in `debrew'
/opt/homebrew/Library/Homebrew/debrew.rb:17:in `install'
/opt/homebrew/Library/Homebrew/build.rb:177:in `block (3 levels) in install'
/opt/homebrew/Library/Homebrew/utils.rb:605:in `with_env'
/opt/homebrew/Library/Homebrew/build.rb:139:in `block (2 levels) in install'
/opt/homebrew/Library/Homebrew/formula.rb:1325:in `block in brew'
/opt/homebrew/Library/Homebrew/formula.rb:2676:in `block (2 levels) in stage'
/opt/homebrew/Library/Homebrew/utils.rb:605:in `with_env'
/opt/homebrew/Library/Homebrew/formula.rb:2675:in `block in stage'
/opt/homebrew/Library/Homebrew/resource.rb:158:in `block (2 levels) in unpack'
/opt/homebrew/Library/Homebrew/download_strategy.rb:116:in `chdir'
/opt/homebrew/Library/Homebrew/download_strategy.rb:116:in `chdir'
/opt/homebrew/Library/Homebrew/download_strategy.rb:103:in `stage'
/opt/homebrew/Library/Homebrew/resource.rb:154:in `block in unpack'
/opt/homebrew/Library/Homebrew/mktemp.rb:77:in `block in run'
/opt/homebrew/Library/Homebrew/mktemp.rb:77:in `chdir'
/opt/homebrew/Library/Homebrew/mktemp.rb:77:in `run'
/opt/homebrew/Library/Homebrew/resource.rb:266:in `stage_resource'
/opt/homebrew/Library/Homebrew/resource.rb:153:in `unpack'
/opt/homebrew/Library/Homebrew/resource.rb:127:in `stage'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/forwardable.rb:230:in `stage'
/opt/homebrew/Library/Homebrew/formula.rb:2655:in `stage'
/opt/homebrew/Library/Homebrew/formula.rb:1318:in `brew'
/opt/homebrew/Library/Homebrew/build.rb:133:in `block in install'
/opt/homebrew/Library/Homebrew/utils.rb:605:in `with_env'
/opt/homebrew/Library/Homebrew/build.rb:128:in `install'
/opt/homebrew/Library/Homebrew/build.rb:229:in `<main>'
```

**Reply by funyx** [reply]
### osx m1

note: the prefix for intel based osx brew install is /usr/local

```bash
$ sudo mkdir /opt/postgis
$ sudo chown $(whoami):admin /opt/postgis
$ cd /opt
$ wget http://postgis.net/stuff/postgis-3.3.3dev.tar.gz
$ tar -xvzf postgis-3.3.3dev.tar.gz -C postgis
$ cd postgis/postgis-3.3.3dev
$ ./configure --with-projdir=/opt/homebrew/opt/proj --with-protobufdir=/opt/homebrew/opt/protobuf-c --with-pgconfig=/opt/homebrew/opt/postgresql@15/bin/pg_config --with-jsondir=/opt/homebrew/opt/json-c "LDFLAGS=$LDFLAGS -L/opt/homebrew/Cellar/gettext/0.21.1/lib" "CFLAGS=-I/opt/homebrew/Cellar/gettext/0.21.1/include"
$ make 
$ make install

psql $
SELECT name, default_version,installed_version FROM pg_available_extensions WHERE name LIKE 'postgis%';
          name          | default_version | installed_version 
------------------------+-----------------+-------------------
 postgis                | 3.3.3dev        | 
 postgis_tiger_geocoder | 3.3.3dev        | 
 postgis_topology       | 3.3.3dev        | 
 postgis_raster         | 3.3.3dev        | 
 postgis_sfcgal         | 3.3.3dev        | 
(6 rows)

```

**Reply by amitsaxena** [reply]
I got the following error when trying to compile from source and running `make`:
```
/usr/local/include/absl/base/policy_checks.h:79:2: error: "C++ versions less than C++14 are not supported."
```
even though my Mac had gcc/g++ version 14 installed. To fix it I had to change the following file in postgis source (after running ./configure): `deps/flatgeobuf/Makefile`. Change `CXX = clang -std=c++11 -x c++` to `CXX = clang -std=c++14 -x c++` and then run `make` and `make install`, and that should hopefully get you out of the rabbit hole I spent countless hours in. Hopefully this helps someone who has been stuck with this and saves you some time. :)

**Reply by mschilde** [reply]
Many thanks @funyx, I just spent the last few hours trying to get Postgis to work on my M1 Mac with Postgresql@15. In the end the only thing that worked was compiling the source with the configure options you provided.

**Reply by prachil007** [reply]
Thank you @funyx for the script. 

## Updated Script for: 
**Postgres 16.2 (homebrew)** 
**Postgis 3.4.2**

```
$ sudo mkdir /opt/postgis
$ sudo chown $(whoami):admin /opt/postgis
$ cd /opt
$ wget http://postgis.net/stuff/postgis-3.4.2.tar.gz
$ tar -xvzf postgis-3.4.2.tar.gz -C postgis
$ cd postgis/postgis-3.4.2
$ ./configure --with-projdir=/opt/homebrew/opt/proj --with-protobufdir=/opt/homebrew/opt/protobuf-c --with-pgconfig=/opt/homebrew/opt/postgresql@16/bin/pg_config --with-jsondir=/opt/homebrew/opt/json-c "LDFLAGS=$LDFLAGS -L/opt/homebrew/Cellar/gettext/0.22.5/lib" "CFLAGS=-I/opt/homebrew/Cellar/gettext/0.22.5/include"
$ make 
$ make install
```

```
psql $
SELECT name, default_version,installed_version FROM pg_available_extensions WHERE name LIKE 'postgis%';
          name          | default_version | installed_version 
------------------------+-----------------+-------------------
 postgis                | 3.4.2        | 
 postgis_tiger_geocoder | 3.4.2        | 
 postgis_topology       | 3.4.2        | 
 postgis_raster         | 3.4.2        | 
 postgis_sfcgal         | 3.4.2        | 
(6 rows)
```

### Enable / Install extention in Postgressql 16

**PostGIS** 

` CREATE EXTENSION postgis;`

**PostGIS Topology** 

` CREATE EXTENSION postgis_topology;`

### **Tips for issues related to dependencies**

1. Install postgressql@16 via homebrew using `brew install postgressql@16`
2. Install postgis via homebrew using `brew install postgis` 
3. Check the version of `gettext` at `/opt/homebrew/Cellar/gettext/` before you proceed and make appropriate changes according to the versions available for `gettext`, `postgressql` and `postgis`.

**Reply by sagadave** [reply]
Trying to build postgis from source using the script above but make fails with an error:

> Built-in generator --c_out specifies a maximum edition PROTO3 which is not the protoc maximum 2023. 
> make[1]: *** [vector_tile.pb-c.c] Error 1

This is on macOS Sonoma 14.6.1 with brew packages postgresql@16 (16.4) and protobuf-c 1.5.0_8 and postgis 3.4.3 from source, but the same error occurs with other recent versions of postgis and postgresql. Does anyone have any insights in to this error? 

make completes successfully if I specify --without-protobuf,  but protobuf is needed of course. 

The same error is also reported at https://stackoverflow.com/questions/77149123/enable-postgis-extension-for-postgresql-16-on-mac-m2, but no solution is described there, sadly.

**Reply by ozeias** [reply]
Try this https://github.com/ozeias/homebrew-postgresql

```shell
brew rm postgresql@16
brew install ozeias/postgresql/postgis@16
```

**Reply by sagadave** [reply]
Thanks, that works great. I see there is a protobuf version update in the homebrew pipeline but not clear yet if it will fix this issue.

**Reply by sagadave** [reply]
The release of protobuf-c 1.5.0_9 appears to have fixed the PROTO3 incompatibility.

**Comment by svoop** [comment]
I'm having the same problem here trying to upgrade our infra to Pg15. The dependency on a specific version of Pg was introduced not long ago:

https://github.com/Homebrew/homebrew-core/commit/69e7ca8d4c354a8d4c652cf53996b496c788279f

@MikeMcQuaid Would you have a minute for some context as to why – and maybe how to work around it? Thanks a bunch!

**Reply by SMillerDev** [reply]
Because we started naming the postgres versions explicitly. It can be easily changed by making a pull request that changes the postgres version.

**Reply by svoop** [reply]
Doesn't this mean in order to use Postgis you are forced to use the Pg Version it is currently pinned to? 

Given Postgis itself accepts any version of Pg, it's very unlikely for everybody using it to always be on the same (latest) version of Pg. Might even be impossible if you cannot update your infra in case your cloud runner (say Render) does not yet support that version.

**Reply by SMillerDev** [reply]
> Doesn't this mean in order to use Postgis you are forced to use the Pg Version it is currently pinned to?

Yes, that is and always has been the case for postgis through Homebrew.

**Reply by svoop** [reply]
IC. Slotting Postgis as well is not really an option since its versions are not in sync with Pg at all. Other distro face the same dilemma, Gentoo allows for more relaxed dependencies like `postgis -> postgresql@9..15`. Is something similar possible for Homebrew as well or are the `@version` postfixes merely conventions with no code parsing them for dependency resolution?

**Reply by MikeMcQuaid** [reply]
@SMillerDev is correct here.

> Is something similar possible for Homebrew as well

No, not really. You'd need to create your own formulae in a tap (`brew tap-new` and `brew extract` may be useful) where you change the version used here. We do not support swapping them in and out.

**Reply by svoop** [reply]
Bummer 😄 I'll try to `brew edit` Postgis to install against Pg15 first. If it works, a PR might follow. Thanks for your help!

**Reply by mschilde** [reply]
Does it make sense to version the Postgis formula then as well (postgis@14 etc) or should it always be hard coded and fixed to the latest Postgres version?

**Comment by vdavez** [comment]
FWIW, @svoop, I'm in the exact same boat. I've tried a bunch of different things to try and get things to make nice, without any luck. If someone figures this out, I'd be grateful!

**Reply by fbonzon** [reply]
> I'm in the exact same boat. I've tried a bunch of different things to try and get things to make nice, without any luck. If someone figures this out, I'd be grateful!

@vdavez I was able to build PostGIS against PostgreSQL 15 (via `brew install --build-from-source postgis`) with the changes from this commit: https://github.com/fbonzon/homebrew-core/commit/05c672d3ee73e63e8d2560f1e3a532dc2729b7b1

Changes are only `postgresql@14` → `postgresql@15` 2x (3x if you include a comment) and `HOMEBREW_PREFIX` → `postgresql.prefix` on line 90.

However, then PostgreSQL will not find the PostGIS installation and raise errors like:
> ERROR:  extension "postgis" is not available
> ERROR:  could not access file "$libdir/postgis-3": No such file or directory

I fixed those errors by adding missing symlinks inside the PostgreSQL directories, pointing to the PostGIS files.

For the first above error:
```shell
cd $(pg_config --sharedir)/extension
find ../../../../../postgis/3.3.2_2/share/postgresql@15/extension -mindepth 1 -maxdepth 1 -print0 | \
  xargs -0 -L 1 ln -s -f
```

For the second above error:
```shell
cd $(pg_config --libdir)/postgresql
ln -s -f /opt/homebrew/Cellar/postgis/3.3.2_2/lib/postgresql/address_standardizer-3.so
ln -s -f /opt/homebrew/Cellar/postgis/3.3.2_2/lib/postgresql/postgis-3.so
ln -s -f /opt/homebrew/Cellar/postgis/3.3.2_2/lib/postgresql/postgis_raster-3.so
ln -s -f /opt/homebrew/Cellar/postgis/3.3.2_2/lib/postgresql/postgis_sfcgal-3.so
ln -s -f /opt/homebrew/Cellar/postgis/3.3.2_2/lib/postgresql/postgis_topology-3.so
```

NOTE:
- `3.3.2_2` is the current version of PostGIS in Homebrew. In case PostGIS is updated in Homebrew, it will need to be upgraded accordingly.
- The above commands needs to be repeated every time a PostGIS or PostgreSQL 15 update happens in Homebrew.
- Adding those missing symlinks manually works for me but seems hacky. There is probably a more elegant solution involving some `brew` commands, but I'm not familiar enough with the Homebrew ecosystem to be able to find it.

**Reply by vdavez** [reply]
Amazing! Thank you! I was able to make it work! I really really appreciate the help!

**Reply by dpritchett** [reply]
Blundering my way through the same issue with `postgresql@13` and `postgis`, thanks for sharing these notes!

I have a feeling the Formula syntax could allow for an install-time option to let a user specify which `postgresql@version-goes-here` package they wanted but I've not customized enough formulae in the past to know where to start.

**Reply by svoop** [reply]
@fbonzon I can only do `brew edit postgis` and then `brew install --build-from-source postgis` when postgis (and therefore postgres@14) is already installed. After uninstalling the two, `brew edit postgis` simply throws "Error: Invalid usage: postgis doesn't exist on disk". (Also, uninstalling only postgresql@14 and keeping the dependency postgis, then `brew edit postgis` and `brew reinstall --build-from-source postgis` doesn't do the trick, the thus installed postgis still references the ole 14 in `postgis/3.3.4_1/share/postgresql@14/extension`. Do you remember how you managed to get around this?

@boris-glumpler Soon, this discussion you've started will celebrate its first anniversary... and unfortunately, we're no bit closer to a real fix than on day one. 😢

**Reply by boris-glumpler** [reply]
I don't expect there to be a solution anymore, tbh. Installing PostGIS from source is fairly easy, tho, so that's what I've been doing.

**Comment by dpritchett** [comment]
I’m on an M2 Mac and I’ve been able to use prebuilt binaries from the
Postgres.app “universal” installer to get pg13+postgis working:
https://postgresapp.com/downloads.html

Haven’t bothered with brew for this one thing but if y’all find a
pg13+postgis formula that worked on ARM Macs I’d likely choose that over
the installer I linked above.

On Tue, Jul 4, 2023 at 10:56 AM Marko Schilde ***@***.***>
wrote:

> Does it make sense to version the Postgis formula then as well ***@***.***
> etc) or should it always be hard coded and fixed to the latest Postgres
> version?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/Homebrew/discussions/3987#discussioncomment-6355080>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABEEDLTI6FCZO3GXBPP2Z3XOQ4MDANCNFSM6AAAAAASL5ERLE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 

___________________________

Daniel Pritchett


📨 ***@***.*** |📱 901-871-0139 | ✍🏻 dpritchett.net | 🐦
@dpritchett
<https://dpritchett.net/>

**Comment by b4stien** [comment]
I've taken a shot at writing a "postgresql@15-postgis" without luck so far. I can't understand how/where the symbolic links are done.

After installing `postgresql@14` / `postgis`, you end up with various symbolic links like `/opt/homebrew/share/postgresql@14/extension/address_standardizer--1.0--3.3.4.sql -> /opt/homebrew/Cellar/postgis/3.3.4_1/share/postgresql@14/extension/address_standardizer--1.0--3.3.4.sql`. As far as I can tell, this links are not created [in the `postgis` formula](https://github.com/Homebrew/homebrew-core/blob/144e7dad8709c06723bbcc0aa57a101c02fb3c59/Formula/p/postgis.rb). Does anyone know where they come from?

I think it's the missing piece for me to make a `postgresql@15-postgis` formula.

**Reply by bahoo** [reply]
Ooh @b4stien — https://postgis.net/docs/manual-2.5/postgis_installation.html#installing_pagc_address_standardizer

> The address_standardizer extension used to be a separate package that required separate download. From PostGIS 2.2 on, it is now bundled in.

**Reply by bahoo** [reply]
Actually that's not that interesting — I think this is closer to what you're looking for ?

https://github.com/postgis/postgis/blob/master/extensions/upgrade-paths-rules.mk#L22-L26

Noting too, that file's been through [a number of changes the last several months](https://github.com/postgis/postgis/commits/master/extensions/upgrade-paths-rules.mk).

**Comment by ozeias** [comment]
I have it working here, and once I have more time, I will do it for PostgreSQL 16.
https://github.com/ozeias/homebrew-postgresql

**Reply by jon-sully** [reply]
Thank you for this! 🙌 your pg16+postgis16 formula worked great for me 💯

**Reply by b4stien** [reply]
I concur: thanks a lot, it works flawlessly.

**Reply by marcmalha** [reply]
Still works. Thank you so much man...❤️
